--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -315,11 +315,11 @@
             compoundtag.putInt("SpawnX", this.respawnPosition.getX());
             compoundtag.putInt("SpawnY", this.respawnPosition.getY());
             compoundtag.putInt("SpawnZ", this.respawnPosition.getZ());
             compoundtag.putBoolean("SpawnForced", this.respawnForced);
             compoundtag.putFloat("SpawnAngle", this.respawnAngle);
-            DataResult dataresult = ResourceLocation.CODEC.encodeStart(NbtOps.INSTANCE, this.respawnDimension.location());
+            DataResult<net.minecraft.nbt.Tag> dataresult = ResourceLocation.CODEC.encodeStart(NbtOps.INSTANCE, this.respawnDimension.location()); // PistonMC - decompile error fix
             Logger logger = ServerPlayer.LOGGER;
 
             logger.getClass();
             dataresult.resultOrPartial(logger::error).ifPresent((tag) -> {
                 compoundtag.put("SpawnDimension", tag);
