--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -671,10 +671,14 @@
 
    public int getMaxPlayers() {
       return this.maxPlayers;
    }
 
+   public void setMaxPlayers(int maxPlayers) { // PistonMC - Add setter
+      this.maxPlayers = maxPlayers;
+   }
+
    public boolean isUsingWhitelist() {
       return this.doWhiteList;
    }
 
    public void setUsingWhiteList(boolean flag) {
