--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -173,10 +173,16 @@
    private int containerCounter;
    public boolean ignoreSlotUpdateHack;
    public int latency;
    public boolean wonGame;
 
+   @Override
+   public org.pistonmc.server.impl.pluginapi.entity.player.PistonServerPlayerEntity getPistonPluginEntityImpl() { // PistonMC - Add pistonEntity getter
+      if(pistonEntity == null) pistonEntity = new org.pistonmc.server.impl.pluginapi.entity.player.PistonServerPlayerEntity(this);
+      return (org.pistonmc.server.impl.pluginapi.entity.player.PistonServerPlayerEntity) pistonEntity;
+   }
+
    public ServerPlayer(MinecraftServer minecraftserver, ServerLevel serverlevel, GameProfile gameprofile, ServerPlayerGameMode serverplayergamemode) {
       super(serverlevel, serverlevel.getSharedSpawnPos(), serverlevel.getSharedSpawnAngle(), gameprofile);
       serverplayergamemode.player = this;
       this.gameMode = serverplayergamemode;
       this.server = minecraftserver;
