--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -625,11 +625,11 @@
         this.getSleepingPos().ifPresent((blockpos) -> {
             compoundtag.putInt("SleepingX", blockpos.getX());
             compoundtag.putInt("SleepingY", blockpos.getY());
             compoundtag.putInt("SleepingZ", blockpos.getZ());
         });
-        DataResult dataresult = this.brain.serializeStart(NbtOps.INSTANCE);
+        DataResult<net.minecraft.nbt.Tag> dataresult = this.brain.serializeStart(NbtOps.INSTANCE); // PistonMC - decompile error fix
         Logger logger = LivingEntity.LOGGER;
 
         logger.getClass();
         dataresult.resultOrPartial(logger::error).ifPresent((net_minecraft_nbt_tag) -> {
             compoundtag.put("Brain", net_minecraft_nbt_tag);
