--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -187,10 +187,13 @@
    private final double[] pistonDeltas = new double[]{0.0D, 0.0D, 0.0D};
    private long pistonDeltasGameTime;
    private EntityDimensions dimensions;
    private float eyeHeight;
 
+   protected org.pistonmc.server.impl.pluginapi.entity.PistonEntity pistonEntity; // PistonMC - Add pistonEntity field
+   public abstract org.pistonmc.server.impl.pluginapi.entity.PistonEntity getPistonPluginEntityImpl(); // PistonMC - Add pistonEntity getter
+
    public Entity(EntityType<?> entitytype, Level level) {
       this.type = entitytype;
       this.level = level;
       this.dimensions = entitytype.getDimensions();
       this.position = Vec3.ZERO;
@@ -375,10 +378,20 @@
    }
 
    public void setPortalCooldown() {
       this.portalCooldown = this.getDimensionChangingDelay();
    }
+   // PistonMC - start
+   private int dimensionChangingDelay = this.getDimensionChangingDelay();
+   public void setDimensionChangingDelay(int dimensionChangingDelay) {
+      if(dimensionChangingDelay < 0) return;
+      this.dimensionChangingDelay = dimensionChangingDelay;
+   }
+   public int getCustomDimensionChangingDelay() {
+      return dimensionChangingDelay;
+   }
+   // PistonMC - end
 
    public boolean isOnPortalCooldown() {
       return this.portalCooldown > 0;
    }
 
@@ -390,10 +403,20 @@
    }
 
    public int getPortalWaitTime() {
       return 0;
    }
+   // PistonMC - start
+   private int portalWaitTime = getPortalWaitTime();
+   public void setPortalWaitTime(int portalWaitTime) {
+      if(portalWaitTime < 0) return;
+      this.portalWaitTime = portalWaitTime;
+   }
+   public int getCustomPortalWaitTime() {
+      return portalWaitTime;
+   }
+   // PistonMC - end
 
    protected void lavaHurt() {
       if (!this.fireImmune()) {
          this.setSecondsOnFire(15);
          this.hurt(DamageSource.LAVA, 4.0F);
@@ -1617,11 +1640,11 @@
       }
    }
 
    protected void handleNetherPortal() {
       if (this.level instanceof ServerLevel) {
-         int i = this.getPortalWaitTime();
+         int i = this.portalWaitTime < 0 ? this.getPortalWaitTime() : this.portalWaitTime; // PistonMC
          ServerLevel serverlevel = (ServerLevel)this.level;
          if (this.isInsidePortal) {
             MinecraftServer minecraftserver = serverlevel.getServer();
             ResourceKey<Level> resourcekey = this.level.dimension() == Level.NETHER ? Level.OVERWORLD : Level.NETHER;
             ServerLevel serverlevel1 = minecraftserver.getLevel(resourcekey);
